core/arch/arm/sta/gprof.c

```diff
@@ -0,0 +1,324 @@
+					  TEE_PARAM_TYPE_NONE);
+	struct tee_ta_session *s;
+	struct sample_buf *sbuf;
+	TEE_Result res;
+	uint32_t len;
+	uint32_t offset;
+	uint32_t scale;
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = get_calling_session(&s);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	len = params[0].value.a;
+	offset = params[0].value.b;
+	scale = params[1].value.a;
+
+	sbuf = calloc(1, sizeof(*sbuf));
+	if (!sbuf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	sbuf->samples = calloc(1, len);
+	if (!sbuf->samples) {
+		free(sbuf);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	sbuf->nsamples = len/sizeof(*sbuf->samples);
+	sbuf->offset = offset;
+	sbuf->scale = scale;
+	sbuf->freq = read_cntfrq();
+	sbuf->enabled = true;
+	s->sbuf = sbuf;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result gprof_stop_pc_sampling(uint32_t param_types,
+					 TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					  TEE_PARAM_TYPE_VALUE_OUTPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+	struct tee_ta_session *s;
+	struct sample_buf *sbuf;
+	uint32_t outsize;
+	TEE_Result res;
+	uint32_t rate;
+	uint32_t size;
+	void *outbuf;
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = get_calling_session(&s);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	sbuf = s->sbuf;
+	if (!sbuf)
+		return TEE_ERROR_BAD_STATE;
+	assert(sbuf->samples);
+
+	/* Stop sampling */
+	if (sbuf->enabled)
+		sbuf->enabled = false;
+
+	rate = ((uint64_t)sbuf->count * sbuf->freq) / sbuf->usr;
+	params[1].value.a = rate;
+
+	size = sbuf->nsamples * sizeof(*sbuf->samples);
+	outbuf = params[0].memref.buffer;
+	outsize = params[0].memref.size;
+	if (!outbuf || !outsize) {
+		params[0].memref.size = size;
+		return TEE_SUCCESS;
+	}
+
+	if (outsize < size)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	memcpy(outbuf, sbuf->samples, size);
+
+	DMSG("TA sampling stats: sample count=%" PRIu32 " user time=%" PRIu64
+	     " cntfrq=%" PRIu32 " rate=%" PRIu32, sbuf->count, sbuf->usr,
+	     sbuf->freq, rate);
+
+	free(sbuf->samples);
+	free(sbuf);
+	s->sbuf = NULL;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result gprof_alloc(uint32_t param_types,
+			      TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					  TEE_PARAM_TYPE_VALUE_OUTPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+	struct tee_ta_session *s;
+	struct user_ta_ctx *utc;
+	tee_mm_entry_t *mm;
+	TEE_Result res;
+	size_t size;
+	paddr_t pa;
+	vaddr_t va;
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = get_calling_session(&s);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	size = params[0].value.a;
+	mm = tee_mm_alloc(&tee_mm_sec_ddr, size);
+	if (!mm)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	utc = to_user_ta_ctx(s->ctx);
+	utc->mm_gprof_heap = mm;
+
+	pa = tee_mm_get_smem(mm);
+	va = tee_mmu_map_gprof_heap(utc, pa, size);
+
+	tee_mmu_set_ctx(&utc->ctx);
```

Comment:
This will shift the virtual addresses of mapped parameters in the current invoke of the TA. The pointers to buffer in the TA arguments will suddenly point into the gprof heap instead. I think we need a better way of doing this.